pkg/nsx/services/ippool/ippool.go

```diff
@@ -0,0 +1,252 @@
+	changed, stale := common.CompareResources(IpAddressPoolBlockSubnetsToComparable(existingIPSubnets), IpAddressPoolBlockSubnetsToComparable(nsxIPSubnets))
+	changedIPSubnets, staleIPSubnets := ComparableToIpAddressPoolBlockSubnets(changed), ComparableToIpAddressPoolBlockSubnets(stale)
+	for i := len(staleIPSubnets) - 1; i >= 0; i-- {
+		staleIPSubnets[i].MarkedForDelete = &MarkedForDelete
+	}
+	finalIPSubnets := append(changedIPSubnets, staleIPSubnets...)
+	if len(finalIPSubnets) > 0 {
+		ipPoolSubnetsUpdated = true
+	}
+
+	if err := service.Operate(nsxIPPool, finalIPSubnets, ipPoolUpdated, ipPoolSubnetsUpdated); err != nil {
+		return false, false, err
+	}
+
+	realizedSubnets, subnetCidrUpdated, e := service.AcquireRealizedSubnetIP(obj)
+	if e != nil {
+		return false, false, e
+	}
+	obj.Status.Subnets = realizedSubnets
+	return subnetCidrUpdated, ipPoolSubnetsUpdated, nil
+}
+
+func (service *IPPoolService) Operate(nsxIPPool *model.IpAddressPool, nsxIPSubnets []*model.IpAddressPoolBlockSubnet, IPPoolUpdated bool, IPPoolSubnetsUpdated bool) error {
+	if !IPPoolUpdated && !IPPoolSubnetsUpdated {
+		return nil
+	}
+	infraIPPool, err := service.WrapHierarchyIPPool(nsxIPPool, nsxIPSubnets)
+	if err != nil {
+		return err
+	}
+	// Get IPPool Type from nsxIPPool
+	IPPoolType := common.IPPoolTypeExternal
+	for _, tag := range nsxIPPool.Tags {
+		if *tag.Scope == common.TagScopeIPPoolCRType {
+			IPPoolType = *tag.Tag
+			break
+		}
+	}
+
+	if IPPoolType == common.IPPoolTypePrivate {
+		ns := service.GetIPPoolNamespace(nsxIPPool)
+		orgProjects := commonctl.ServiceMediator.GetOrgProjectVPC(ns)
+		if len(orgProjects) == 0 {
+			err = util.NoEffectiveOption{Desc: "no effective org and project for ippool"}
+		} else {
+			err = service.NSXClient.ProjectInfraClient.Patch(orgProjects[0].OrgID, orgProjects[0].ProjectID, *infraIPPool,
+				&EnforceRevisionCheckParam)
+		}
+	} else if IPPoolType == common.IPPoolTypeExternal {
+		err = service.NSXClient.InfraClient.Patch(*infraIPPool, &EnforceRevisionCheckParam)
+	} else {
+		err = util.NoEffectiveOption{Desc: "no effective ippool type"}
+	}
+	if err != nil {
+		return err
+	}
+	if IPPoolUpdated {
+		err = service.ipPoolStore.Operate(nsxIPPool)
+		if err != nil {
+			return err
+		}
+	}
+	if IPPoolSubnetsUpdated {
+		err = service.ipPoolBlockSubnetStore.Operate(nsxIPSubnets)
+		if err != nil {
+			return err
+		}
+	}
+	log.V(1).Info("successfully created or updated ippool and ip subnets", "nsxIPPool", nsxIPPool)
+	return nil
+}
+
+func (service *IPPoolService) AcquireRealizedSubnetIP(obj *v1alpha2.IPPool) ([]v1alpha2.SubnetResult, bool, error) {
+	var realizedSubnets []v1alpha2.SubnetResult
+	subnetCidrUpdated := false
+	for _, subnetRequest := range obj.Spec.Subnets {
+		// check if the subnet is already realized
+		realized := false
+		realizedSubnet := v1alpha2.SubnetResult{Name: subnetRequest.Name}
+		for _, statusSubnet := range obj.Status.Subnets {
```

Comment:
Shall we also consider the scenario that some subnets in status are not needed anymore?  meaning there are no such subnets in SubnetRequest.

Reply:
Actually considered. obj.Spec.Subnets are only the latest subnets required, the stale subnets are ignored before.

Reply:
If there are some realized subnets in status, but  the subnet is deleted in obj.Spec.Subnets,  can the realized subnet in status be deleted from NSX?

Reply:
Yes, it can.